include/picongpu/fields/MaxwellSolver/ArbitraryOrderFDTD/Derivative.hpp

```diff
@@ -0,0 +1,234 @@
+            // upperIndex: 1 if ( Forward ) else 0
+            upperIndex[ T_direction ] = static_cast<int32_t>(T_neighbors) - static_cast<int32_t>(T_lowerNeighbors);
+
+            AOFDTDWeights< T_neighbors > const weights{};
+
+            // shortest distance finite difference as initial value
+            auto finiteDifference = weights[0] *
+                ( data( upperIndex ) - data( lowerIndex ) );
+
+            // Compute next finite differences according to order
+            for ( uint32_t l = 1; l < T_neighbors; ++l )
+            {
+                lowerIndex[ T_direction ] -= 1;
+                upperIndex[ T_direction ] += 1;
+
+                finiteDifference += weights[l] *
+                    ( data( upperIndex ) - data( lowerIndex ) );
+            }
+
+            return finiteDifference / cellSize[ T_direction ];
+        }
+    };
+} // namespace detail
+
+
+    /**@{*/
+    /** Functors for forward and backward derivative along the given direction used in ArbitraryOrderFDTD solver
+     *
+     * Compute an approximation of the derivative of a field f by a finite difference of
+     * order 2 * T_neighbors, where T_neighbors is the number of neighbors
+     * used to calculate the finite difference.
+     *
+     * This finite difference approximations for the forward and backward derivative are computed on a staggered grid.
+     * That is, the forward derivative will be known at a position i+1/2, if the field f is known
+     * at 2 * T_neighbors grid nodes i - T_neighbors + 1, i - T_neighbors + 2, ..., i + T_neighbors.
+     * The backward derivative will be known at a position i-1/2, if the field f is known
+     * at 2 * T_neighbors grid nodes i - T_neighbors, i - T_neighbors + 1, ..., i + T_neighbors - 1.
+     *
+     * The finite difference calculation can be expressed as a sum of finite differences where the
+     * distance of field components used in individual finite differences computations increases, e.g.
+     *     D_x f(i+1/2) = sum_{l=0}^{T_neighbors-1} g_l^{2T_neighbors} * ( f(i+1+l) - f(i-l) ) / dx,
+     * for the forward derivative and where D_x is the derivative operator along x, dx the grid spacing along x, and
+     * g_l^{2T_neighbors} weightings for the finite differences of different distance l from the point i
+     * of computation.
+     *
+     * @tparam T_neighbors Number of neighbors used to calculate
+     *                     the derivative from finite differences.
+     *                     Order of derivative approximation is
+     *                     2 * T_neighbors
+     *
+     * @tparam T_direction direction to take derivative in, 0 = x, 1 = y, 2 = z
+     */
+     template<
+        uint32_t T_neighbors,
+        uint32_t T_direction
+     >
+     using ForwardDerivativeFunctor = detail::GeneralAofdtdDerivative< T_neighbors - 1, T_neighbors, T_neighbors, T_direction >;
```

Comment:
These template parameters are better formatted on separate lines, as there are a few, and the line is long.